api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,656 @@
+	minCapacity := int(aws.Int64Value(asg.MinSize))
+	if scale < minCapacity {
+		log.Printf("Scale %d is below the minimum capacity of %d. Setting desired capacity to %d.", scale, minCapacity, minCapacity)
+		scale = minCapacity
+	}
+
+	currentCapacity := int(aws.Int64Value(asg.DesiredCapacity))
+	if scale == currentCapacity {
+		log.Printf("Environment %s is at desired scale of %d. No scaling action required.", ecsEnvironmentID, scale)
+		return scale, nil
+	}
+
+	if scale < currentCapacity {
+		input := &autoscaling.SetDesiredCapacityInput{}
+		input.SetAutoScalingGroupName(aws.StringValue(asg.AutoScalingGroupName))
+		input.SetDesiredCapacity(int64(scale))
+
+		if _, err := e.Client.AutoScaling.SetDesiredCapacity(input); err != nil {
+			return 0, err
+		}
+	}
+
+	// choose which instances to terminate during our scale down process
+	// instead of having asg randomly selecting instances
+	// e.g. if we scale from 5->3, we can terminate up to 2 unused instances
+	maxNumberOfInstancesToTerminate := currentCapacity - scale
+
+	canTerminate := func(i int) bool {
+		if i+1 > maxNumberOfInstancesToTerminate {
+			return false
+		}
+
+		if i > len(unusedProviders)-1 {
+			return false
+		}
+
+		return true
+	}
+
+	for i := 0; canTerminate(i); i++ {
+		unusedProvider := unusedProviders[i]
+		log.Printf("Environment %s terminating unused instance '%s'", ecsEnvironmentID, unusedProvider.ID)
+
+		input := &autoscaling.TerminateInstanceInAutoScalingGroupInput{}
+		input.SetInstanceId(unusedProvider.ID)
+		input.SetShouldDecrementDesiredCapacity(false)
+
+		if _, err := e.Client.AutoScaling.TerminateInstanceInAutoScalingGroup(input); err != nil {
+			return 0, err
+		}
+	}
+
+	return scale, nil
+}
+
+func (e *EnvironmentScaler) determineSortPriority(providers []*models.Resource, consumers []*models.Resource, clusterName string) int {
+	cpuInstanceSize := len(providers)
+	memoryInstanceSize := len(providers)
+
+	for sortMethod := range []int{CPU, Memory} {
+
+		for _, consumer := range consumers {
+			hasRoom := false
+
+			switch sortMethod {
+			case CPU:
+				sortByCPU(providers, consumers)
+			case Memory:
+				sortByMemory(providers, consumers)
+			}
+
+			// next, place any unused providers in the back of the list
+			// that way, we can can delete them if we avoid placing any tasks in them
+			sortByUsage(providers)
+
+			for _, provider := range providers {
+				if hasResourcesFor(*consumer, *provider) {
+					hasRoom = true
+					subtractResourcesFor(consumer, provider)
+					break
+				}
+			}
+
+			if !hasRoom {
+				newProvider, _ := e.calculateNewProvider(clusterName)
+				if !hasResourcesFor(*consumer, *newProvider) {
+					continue
+				}
+
+				switch sortMethod {
+				case CPU:
+					cpuInstanceSize++
+				case Memory:
+					memoryInstanceSize++
+				}
+			}
+		}
+	}
+
+	if cpuInstanceSize > memoryInstanceSize {
+		return CPU
+	}
+
+	return Memory
+}
+
+func hasResourcesFor(consumer models.Resource, provider models.Resource) bool {
```

Comment:
needs a better function name

Reply:
these functions already exist on the resource consumer - why are they duplicated here?

Reply:
FWIW, I didn't see these sorts of functions ever defined as methods on a `ResourceProvider` or `ResourceConsumer`. I only ever saw these functions as standalone helper functions that took a provider and a consumer, and it was a little opaque in regards to what was happening without reading through the function itself.

The changes in #440 have refactored `HasResourcesFor` and `SubtractResourcesFor` into methods on `ResourceProvider`.